SUP-1488: CircleCI docs (#41)
buildkite/migration: docs/CircleCI.md

# CircleCI

The Buildkite Migration tool's currently supported, partially supported and unsupported properties in translation CircleCI pipelines to Buildkite pipelines are listed below:

## Supported properties

### aliases

An `alias` is a definition of a reusable snippet of a CircleCI pipeline. Defined with a `&` (anchor) - these are then substituted into the required section of the pipeline with `*`: for example, `*tests`.

The Buildkite Migration tool supports reading configuration defined from anchors and utilised through the use of an `alias`.

### commands

> A `command` defined in a CircleCI pipeline is a reusable set of instructions with parameters that can be inserted into required `job` executions. Commands can have their own set list of `steps` that are translated through to the generated [command step](https://buildkite.com/docs/pipelines/command-step)'s `commands`. If a `command` contains a `parameters` key, its default parameters that is specified within the block is passed through into the specific command if it is utilised.

Reviewer: `parameters that is specified`

Should be `are specified`.

Author: Tweaked

> A `command` defined in a CircleCI pipeline is a reusable set of instructions with parameters that can be inserted into required `job` executions. Commands can have their own set list of `steps` that are translated through to the generated [command step](https://buildkite.com/docs/pipelines/command-step)'s `commands`. If a `command` contains a `parameters` key, its default parameters that are specified within the block is passed through into the specific command if it is utilised.

Reviewer 2: I think that the parameters part is written in a confusing way. Maybe something like:

```suggestion
A `command` defined in a CircleCI pipeline is a reusable set of instructions with parameters that can be inserted into required `job` executions. Commands can have their own set list of `steps` that are translated through to the generated [command step](https://buildkite.com/docs/pipelines/command-step)'s `commands`. If a `command` contains a `parameters` key, they are respected when used in jobs/workflows and their defaults values used when not specified.
```

Author: Reworded - I think that sounds better!

### workflows

A CircleCI `workflow` is a collection of `jobs`, whose ordering defines how a CircleCI pipeline is run. Workflows can be specified to run its collection of `jobs` sequentially, or in parallel, and allows for specifying dependencies (requirements) between each `workflow` that is configured. 

> Each `job` that is defined as part of a `workflow` within a CircleCI pipeline will be transitioned to a Buildkite [command step](https://buildkite.com/docs/pipelines/command-step) within the generated pipeline.

Reviewer 2: You should also clarify that workflows will be translated to group steps in the resulting pipeline

Author: Added 👆

## Partially supported properties 

### jobs

CircleCI `jobs` are a collection of steps that are run on a single worker unit: whether that is on a host directly, or on a virtualised host (such as within a Docker container). 

The Buildkite Migration tool currently supports the following keys relative to CircleCI `jobs`:

#### General

- `environment`: The `job` level environment variables of a CircleCI pipeline. Applied in the generated [command step](https://buildkite.com/docs/pipelines/command-step) as [step level](https://buildkite.com/docs/pipelines/environment-variables#runtime-variable-interpolation) environment variables with the `env` key.
- `parameters`: Reusable keys that are used within `step` definitions within a `job`. Default parameters that are specified in a `parameters` block are passed through into the [command step](https://buildkite.com/docs/pipelines/command-step)'s `commands` if specified.

> - `steps`: A collection of commands that are executed as part of a CircleCI `job`. Steps can be defined within an `alias`.

Reviewer 2: you should specify that all steps in a job will be commands of a singular step in a buildkite pipeline to ensure they share an execution environment.

Author: Added

> - `steps`: A collection of commands that are executed as part of a CircleCI `job`. Steps can be defined within an `alias`.

Reviewer 2: You should also create a section here with CircleCI builtin steps as there are a few that are also not supported/translatable

Author: Added the others - specifically ones like `shell` (a Buildkite agent start option), `parallelism` (a option we support in steps), `branches` (workflow support for build conditionals) etc

#### Docker

> - `docker`: Specifies that the `job` will run within a Docker container (by its `image` property). Translates to the use of the [Docker Buildkite Plugin](https://github.com/buildkite-plugins/docker-buildkite-plugin) with the specified image defined - and additionally: the [Docker Login Plugin](https://github.com/buildkite-plugins/docker-login-buildkite-plugin) and the [ECR Buildkite Plugin](https://github.com/buildkite-plugins/ecr-buildkite-plugin) if `auth` or `aws-auth` is specified in the `docker` parameter respectfully. The [Docker Compose Buildkite Plugin](https://github.com/buildkite-plugins/docker-compose-buildkite-plugin) will need to be used for multi-container `jobs`. Sets the [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) for the generated [command step](https://buildkite.com/docs/pipelines/command-step) to `executor_type: docker`.

Reviewer: `is specified in the `docker` parameter respectfully`

I'm not sure if this reads correctly?

Reviewer: This feels wordier than it needs to be.

Author: Shortened 👍 (multi-container definitions are mentioned in a translation)

- `resource_class`: The specification of compute that the executor will require in running a job. This is used to specify the `resource_class` [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tag for the corresponding [command step](https://buildkite.com/docs/pipelines/command-step).

#### macOS 

- `macos`: Specifies that the `job` will run on a macOS bases execution environment. The [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tags for the generated [command step](https://buildkite.com/docs/pipelines/command-step) will be set to `executor_type: osx`, as well as the specified version of `xcode` from the `macos` parameters as `executor_xcode: <version>`.

#### Windows

- `windows`: Specifies that the `job` will run on a Windows based execution environment. The [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tags for the generated [command step](https://buildkite.com/docs/pipelines/command-step) will be set to `executor_type: windows`.

Currently, these following `job` parameters are not supported by the Buildkite Migration tool:

> - `executors`: The `executor` key is currently not supported yet (Supported keys for `job` execution environments are `docker`, `machine`, `macos` and `windows`).

Reviewer 2: This is quite confusing as you just specified how each of the values would be translated.

The reality is that there is no translation for this key because of the fact that the execution environment in buildkite is defined by the agent it will run on. To allow for different configurations, agent tags are used (but queues are also available).

Author: I revised this in the latest commit - to more state that the actual ex environment is determined by a combination of queue/tag config that is setup for specific agents

- `working_directory`: The location of the executor where steps are run in. If set, a change directory (`cd`) command is created within the shared `commands` of a Buildkite [command step](https://buildkite.com/docs/pipelines/command-step) to the desired location.

#### Docker (Execution Environment)

- `docker`: Specifies that the `job` will run within a Docker container (by its `image` property) with the use of the [Docker Buildkite Plugin](https://github.com/buildkite-plugins/docker-buildkite-plugin). Additionally, the [Docker Login Plugin](https://github.com/buildkite-plugins/docker-login-buildkite-plugin) is appended if an `auth` property is defined, or the [ECR Buildkite Plugin](https://github.com/buildkite-plugins/ecr-buildkite-plugin) if an `aws-auth` property is defined within the `docker` property. Sets the [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) for the generated [command step](https://buildkite.com/docs/pipelines/command-step) to `executor_type: docker`.

#### Machine (Execution Environment)

> - `machine`: Specifies that the `job` will run on a machine execution environment. The `resource_class` parameter will be utilised to specify [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) for the generated [command step](https://buildkite.com/docs/pipelines/command-step), in addition to setting the additional tags of ``executor_type: machine` and `executor_image: self-hosted`.

Reviewer 2: the `resource_class` part is mentioned immediately afterwards, so this could be simplified to:
```suggestion
- `machine`: Specifies that the `job` will run on a machine execution environment. This translates to [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) for the generated [command step](https://buildkite.com/docs/pipelines/command-step) through the  tags `executor_type: machine` and `executor_image: self-hosted`.
```

Author: Commited 👍

- `resource_class`: The specification of compute that the executor will require in running a job. This is used to specify the `resource_class` [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tag for the corresponding [command step](https://buildkite.com/docs/pipelines/command-step).

#### macOS (Execution Environment)

- `macos`: Specifies that the `job` will run on a macOS bases execution environment. The [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tags for the generated [command step](https://buildkite.com/docs/pipelines/command-step) will be set to `executor_type: osx`, as well as the specified version of `xcode` from the `macos` parameters as `executor_xcode: <version>`.

#### Windows (Execution Environment)

- `windows`: Specifies that the `job` will run on a Windows based execution environment. The [agent targeting](https://buildkite.com/docs/pipelines/defining-steps#targeting-specific-agents) tags for the generated [command step](https://buildkite.com/docs/pipelines/command-step) will be set to `executor_type: windows`.

> Currently, these following `job` parameters are not supported by the Buildkite Migration tool:
> 
> #### General

Reviewer 2: I would make this a whole section:

```suggestion
#### Unsupported properties

The following `job` parameters are currently not supported by the Buildkite Migration tool:
```

Author: In the latest commit :)
